R/utils-latency.R

```diff
@@ -135,52 +136,98 @@ set_asof <- function(new_data, info) {
     cli::cli_abort(paste(
-      "the data matrix `as_of` value is {as_of}, ",
+      "the data matrix `forecast_date` value is {forecast_date}, ",
       "and not a valid `time_type` with type ",
       "matching `time_value`'s type of ",
-      "{typeof(new_data$time_value)}."
+      "{class(max_time)}."
     ))
   }
-  if (is.null(as_of) || is.na(as_of)) {
+  if (is.null(forecast_date) || is.na(forecast_date)) {
     cli::cli_warn(paste(
-      "epi_data's `as_of` was {as_of}, setting to ",
+      "epi_data's `forecast_date` was {forecast_date}, setting to ",
       "the latest time value, {max_time}."
     ))
-    as_of <- max_time
-  } else if (as_of < max_time) {
+    forecast_date <- max_time
+  } else if (forecast_date < max_time) {
     cli::cli_abort(paste(
-      "`as_of` ({(as_of)}) is before the most ",
+      "`forecast_date` ({(forecast_date)}) is before the most ",
       "recent data ({max_time}). Remove before ",
       "predicting."
     ))
   }
   # TODO cover the rest of the possible types for as_of and max_time...
-  if (class(time_values) == "Date") {
-    as_of <- as.Date(as_of)
+  if (class(max_time) == "Date") {
+    forecast_date <- as.Date(forecast_date)
   }
-  return(as_of)
+  return(forecast_date)
 }
 
 #' the latency is also the amount the shift is off by
 #' @param sign_shift integer. 1 if lag and -1 if ahead. These represent how you
 #'   need to shift the data to bring the 3 day lagged value to today.
 #' @keywords internal
-get_latency <- function(new_data, as_of, column, shift_amount, sign_shift) {
+get_latency <- function(new_data, forecast_date, column, shift_amount, sign_shift, epi_keys_checked) {
   shift_max_date <- new_data %>%
     drop_na(all_of(column)) %>%
+    {
+      # null and "" don't work in `group_by`
+      if (!is.null(epi_keys_checked) && epi_keys_checked != "") {
+        group_by(., get(epi_keys_checked))
+      } else {
+        .
+      }
+    } %>%
+    summarize(time_value = max(time_value)) %>%
     pull(time_value) %>%
-    max()
-  return(as.integer(sign_shift * (as_of - shift_max_date) + shift_amount))
+    min()
+  return(as.integer(sign_shift * (as.Date(forecast_date) - shift_max_date) + shift_amount))
+}
+
+
+
+#' get the target date while in a layer
+get_forecast_date_in_layer <- function(this_recipe, workflow_max_time_value, new_data) {
+  max_time_value <- max(
+    workflow_max_time_value,
+    this_recipe$max_time_value,
+    max(new_data$time_value)
+  )
+  if (this_recipe %>% recipes::detect_step("adjust_latency")) {
+    # get the as_of in an `adjust_latency` step, regardless of where
+    handpicked_as_of <- map(
+      this_recipe$steps,
+      function(x) {
+        if (inherits(this_recipe$steps[[3]], "step_adjust_latency")) x$as_of
+      }
+    ) %>% Filter(Negate(is.null), .)
+    if (length(handpicked_as_of) > 0) {
+      max_time_value <- handpicked_as_of[[1]]
+    } else {
```

Comment:
the `this_recipe$steps[[3]]` was left over from development, surprised it didn't cause any errors yet.

`handpicked_as_of` _should_ have only value (otherwise they have multiple `step_adjust_latency`s, which shouldn't work). I suppose I should add a check that there's only one `step_adjust_latency` during step creation